### PR TITLE
Feature/get stormtype from ibtracs

### PR DESCRIPTION
**Pull request by ChrisFairless**
Changes proposed in this PR:
- When creating a trackset from IBTrACS also read the 'Nature' of a storm. The nature gives information about whether a system is a disturbance, tropical storm, post-transition extratropical storm, or subtropical storm (or whether agencies disagree or don't report). This is potentially useful in visualisations and for generating synthetic tracks.

**Notes:**
- Not urgent! We don't need this in time for the next release!
- Do we need a discussion? There are lots of variables that we exclude when reading from IBTrACS and we can't include them all. A more flexible alternative would be to add a keyword for 'other columns' you want from the file.
- The 'nature' variable is only explicitly created for tracksets from current IBTrACS (not olders csvs), and not for CHAZ, STORM, etc. It looks like there are already other properties that are IBTrACS-only, e.g radius of outermost closed isobar, so hopefully this is ok. I added a check for existence of the variable where relevant.
- When interpolating with `equal_timestep` I chose a forward fill method, rather than the default nearest neighbour.

### PR Author Checklist

- [x] Read the [Contribution Guide][contrib]
- [x] Correct target branch selected (if unsure, select `develop`)
- [x] Source branch up-to-date with target branch
- [x] [Documentation](https://climada-python.readthedocs.io/en/latest/guide/Guide_PythonDos-n-Donts.html#2.--Commenting-&-Documenting) updated
- [x] [Tests][testing] updated
- [x] [Tests][testing] passing
- [x] No new [linter issues][linter]
- [ ] [Changelog](https://github.com/CLIMADA-project/climada_python/blob/main/CHANGELOG.md) updated

### PR Reviewer Checklist

- [ ] Read the [Contribution Guide][contrib]
- [ ] [CLIMADA Reviewer Checklist](https://climada-python.readthedocs.io/en/latest/guide/Guide_Reviewer_Checklist.html) passed
- [ ] [Tests][testing] passing
- [ ] No new [linter issues][linter]

[contrib]: https://github.com/CLIMADA-project/climada_python/blob/main/CONTRIBUTING.md
[testing]: https://climada-python.readthedocs.io/en/latest/guide/Guide_Continuous_Integration_and_Testing.html
[linter]: https://climada-python.readthedocs.io/en/stable/guide/Guide_Continuous_Integration_and_Testing.html#3.C.--Static-Code-Analysis

### CODE REVIEW

**Comment by tovogt** [comment]
Thanks for the PR! I can imagine that this might be useful information for some people.


> * Do we need a discussion? There are lots of variables that we exclude when reading from IBTrACS and we can't include them all. A more flexible alternative would be to add a keyword for 'other columns' you want from the file.

Here is a list of all non-agency-specific variables that we currently ignore:
```python
[
    'numobs', 'season', 'number', 'subbasin', 'name', 'source_usa', 'source_jma',
    'source_cma', 'source_hko', 'source_new', 'source_reu', 'source_bom', 'source_nad',
    'source_wel', 'source_td5', 'source_td6', 'source_ds8', 'source_neu', 'source_mlc',
    'iso_time', 'nature', 'lat', 'lon', 'wmo_wind', 'wmo_pres', 'wmo_agency', 'track_type',
    'main_track_sid', 'dist2land', 'landfall', 'iflag', 'storm_speed', 'storm_dir'
]
```
When including agency-specific variables, the list is even longer. I tend to prefer supporting a keyword argument for "other variables" to include instead of adding explicit support for the "nature" variable. Maybe we can start by supporting non-agency-specific variables, because the agency-specific need to be treated a bit differently?

> * The 'nature' variable is only explicitly created for tracksets from current IBTrACS (not olders csvs), and not for CHAZ, STORM, etc. It looks like there are already other properties that are IBTrACS-only, e.g radius of outermost closed isobar, so hopefully this is ok. I added a check for existence of the variable where relevant.

In general, I think adding additional variables should be fine. We have to avoid breaking file read/write operations though.

> * When interpolating with `equal_timestep` I chose a forward fill method, rather than the default nearest neighbour.

Why would you do that?

File: climada/hazard/tc_tracks.py [hunk 1]
```diff
@@ -1885,6 +1890,8 @@ def _read_ibtracs_csv_single(file_name):
     tr_ds.attrs['sid'] = name
     tr_ds.attrs['orig_event_flag'] = bool(dfr['original_data']. values[0])
     tr_ds.attrs['data_provider'] = dfr['data_provider'].values[0]
+    if hasattr(dfr, 'nature'):
+        tr_ds['nature'] = ('time', dfr['nature'].values.astype('<U2'))
```
Comment (tovogt) [inline climada/hazard/tc_tracks.py:1894]: I don't think it's necessary to support this. The legacy CSV format is a completely undocumented data type. I have never seen such a file apart from our test file. I don't know anyone who is using this functionality.

Comment (ChrisFairless) [reply climada/hazard/tc_tracks.py:1894]: Good point! This was just to stop the tests failing :p

**Comment by ChrisFairless** [comment]
Sorry, I missed your main response last week!

> When including agency-specific variables, the list is even longer. I tend to prefer supporting a keyword argument for "other variables" to include instead of adding explicit support for the "nature" variable. Maybe we can start by supporting non-agency-specific variables, because the agency-specific need to be treated a bit differently?

I think this makes sense. I'll update the pull request, but for non-agency-specific variables. A few of the variables you list will be useful in different situations.

> In general, I think adding additional variables should be fine. We have to avoid breaking file read/write operations though.

To be clear: are you suggesting always reading 'nature' _and_ allowing additional variables, or just allowing additional variables here?

> * When interpolating with `equal_timestep` I chose a forward fill method, rather than the default nearest neighbour.

Not sure why I chose this now ... I think it was from an overabundance of caution around landfalls. Let's go for nearest neighbour then.

**Comment by tovogt** [comment]
> > In general, I think adding additional variables should be fine. We have to avoid breaking file read/write operations though.
> 
> To be clear: are you suggesting always reading 'nature' _and_ allowing additional variables, or just allowing additional variables here?

I think this remark is mostly about this section of code:
https://github.com/CLIMADA-project/climada_python/blob/cfebf28ab6a4fe1bed5938e530f959e54db5eb40/climada/hazard/tc_tracks.py#L1387-L1389
It's a bit unclear what to do when we support additional variables of varying dtype. We can't just assume that all "object" variables are actually supposed to be `"<U2"` like here:
```python
# when writing '<U2' and reading in again, xarray reads as dtype 'object'. undo this:
for varname in track.variables:
    if track[varname].dtype == "object":
        track[varname] = track[varname].astype('<U2')
```
Maybe we can at least assume that all "object" dtypes are actually strings. Then we could do something like this:
```python
# when writing '<U2' and reading in again, xarray reads as dtype 'object'. undo this:
for varname in track.variables:
    if track[varname].dtype == "object":
        strlen = track[varname].str.len().max().item()
        track[varname] = track[varname].astype(f"<U{strlen}")
```

File: climada/hazard/tc_tracks.py (hunk 1 above)
Comment (tovogt) [reply climada/hazard/tc_tracks.py:1894]: I don't understand why the tests would fail without this. Can you elaborate, please?

**Comment by ChrisFairless** [comment]
Assuming a string seems reasonable. Is there a reason we always want to limit its length? What are the memory/efficiency gains there?

**Comment by tovogt** [comment]
The most important aspect for me was that you should be able to cycle, i.e., storing data and reading it again should yield the identical object. I don't know about the memory requirements.

File: climada/hazard/tc_tracks.py (hunk 1 above)
Comment (ChrisFairless) [reply climada/hazard/tc_tracks.py:1894]: Oh lol the tests were failing due to a line I added and later removed. We can definitely drop this.
I'll put together a new pull request shortly.

**Comment by peanutfun** [comment]
> I'll put together a new pull request shortly.

@ChrisFairless I don't think that's necessary, you can simply rewrite your changes here. But if you still want to, I would wait with my review until this or the new PR are ready. In general, I second @tovogt's suggesion of reading any additional variable through keyword arguments.

As a side note, I guess this discussion goes to show that it's helpful to discuss any additions *before* implementing them, although I very much appreciate the initiative of coming up with suggestions through a PR in the first place 😅

**Comment by tovogt** [comment]
> As a side note, I guess this discussion goes to show that it's helpful to discuss any additions _before_ implementing them, although I very much appreciate the initiative of coming up with suggestions through a PR in the first place sweat_smile

In my opinion, it's good to create a draft PR right from the start when proposing a new feature, and have a discussion in the PR. I know that it somehow makes the PR hard to follow because it has so many posts. Because of that, it might make sense to create a new PR at some point. But the big advantage is that you avoid a lot of misunderstandings because you talk directly about concrete code changes. When talking about features at a very high level, I often discuss for a while before noticing that there was actually a misunderstanding about the kind of change we are actually discussing.

**Comment by ChrisFairless** [comment]
Definitely definitely agree here!!

But in this case I needed the feature pretty urgently so I coded first, PR'd later 😅

**Comment by peanutfun** [comment]
How to continue here? There seems to be at least one unresolved discussion. @ChrisFairless will you update this PR or should we close it in favor of a new one?

**Comment by ChrisFairless** [comment]
Fine to close this.

I want to update things as discussed above, but it'll take some time I don't have right now.

**Comment by bguillod** [comment]
Since nobody seems to have capacity to implement the flexible reading of any variable in IBTrACS, it looks to me like the requested addition to this PR is not very relevant or does not have any specific use case for the moment.
On the other hand, we urgently need the storm type information (and only that additional variable) for the new TC synthetic tracks methodology. Therefore I'd like this PR to be reconsidered for merging into the develop branch, to allow us to move ahead and finalize the new synthetic tracks model. Does this make sense?

**Comment by chahank** [comment]
Please address the review comments in the code and then we can proceed. Thanks.

**Comment by tovogt** [comment]
If nobody has time to do this "right", I will take this on immediately. We have discussed everything, the coding should be really easy. I don't see why we would go with the "nature" variable only. I will report back later today.

**Comment by bguillod** [comment]
> If nobody has time to do this "right", I will take this on immediately. We have discussed everything, the coding should be really easy. I don't see why we would go with the "nature" variable only. I will report back later today.

Sure, you're most welcome to implement it if you think that's a worthwhile addition. Thanks a lot, very much appreciated!

**Comment by emanuel-schmid** [comment]
🙌

**Comment by tovogt** [comment]
Let's continue here: #728
